djautotask/syncrest.py

```diff
@@ -0,0 +1,402 @@
+            # same ticket at once.
+            msg = "IntegrityError while attempting to create {}." \
+                  " Error: {}".format(self.model_class, e)
+            logger.error(msg)
+            raise InvalidObjectException(msg)
+
+        if result == CREATED:
+            result_log = 'Created'
+        elif result == UPDATED:
+            result_log = 'Updated'
+        else:
+            result_log = 'Skipped'
+
+        logger.info('{}: {} {}'.format(
+            result_log,
+            self.model_class.__bases__[0].__name__,
+            instance
+        ))
+
+        return instance, result
+
+    def prune_stale_records(self, initial_ids, synced_ids):
+        """
+        Delete records that existed when sync started but were
+        not seen as we iterated through all records from REST API.
+        """
+        stale_ids = initial_ids - synced_ids
+        deleted_count = 0
+        if stale_ids:
+            delete_qset = self.get_delete_qset(stale_ids)
+            deleted_count = delete_qset.count()
+
+            pre_delete_result = None
+            if self.pre_delete_callback:
+                pre_delete_result = self.pre_delete_callback(
+                    *self.pre_delete_args
+                )
+            logger.info(
+                'Removing {} stale records for model: {}'.format(
+                    len(stale_ids), self.model_class.__bases__[0].__name__,
+                )
+            )
+            delete_qset.delete()
+            if self.post_delete_callback:
+                self.post_delete_callback(pre_delete_result)
+
+        return deleted_count
+
+    def get_delete_qset(self, stale_ids):
+        return self.model_class.objects.filter(pk__in=stale_ids)
+
+    def get_sync_job_qset(self):
+        return models.SyncJob.objects.filter(
+            entity_name=self.model_class.__bases__[0].__name__
+        )
+
+    @log_sync_job
+    def sync(self):
+        sync_job_qset = self.get_sync_job_qset()
+
+        # Since the job is created before it begins, make sure to exclude
+        # itself, and at least one other sync job exists.
+        if sync_job_qset.count() > 1 and not self.full:
+            last_sync_job_time = sync_job_qset.exclude(
+                id=sync_job_qset.last().id).last().start_time.isoformat()
+            self.api_conditions.append(
+                "lastUpdated>[{0}]".format(last_sync_job_time)
+            )
+        results = SyncResults()
+        results = self.get(results)
+
+        if self.full:
+            # Set of IDs of all records prior to sync,
+            # to find stale records for deletion.
+            initial_ids = self._instance_ids()
+
+            results.deleted_count = self.prune_stale_records(
+                initial_ids, results.synced_ids
+            )
+
+        return results.created_count, results.updated_count, \
+            results.skipped_count, results.deleted_count
+
+    def callback_sync(self, filter_params):
```

Comment:
I think this method can be removed for now. We will probably need something similar later on when we move handling of webhook/callbacks to the REST API but no need to add now.

Reply:
Hi Cameron, regarding pagination in AT RESTAPI, it is a little different with that in CW. In CW, we know the exact address of the 2nd page, and we can call it directly, which is the same as access array. But in AT, it is something like iterator (bi-directional). i.e. we don't know the address of 2nd page of the API. When we got the result of the API, it contains the addresses in neighbourhood such as previous url and next url.

Reply:
It looks something like:

    "pageDetails": {
        "count": 3,
        "requestCount": 3,
        "prevPageUrl": null,
        "nextPageUrl": "https://webservices2.autotask.net/atservicesrest/v1.0/Contacts/query/next?paging=%7b%22pageSize%22%3a3%2c%22previousIds%22%3a%5b29683585%5d%2c%22nextIds%22%3a%5b29683588%5d%7d"
    }

Therefore, it has no direct access feature of some page, and it has different structure as it is.

Reply:
I commented too rapidly after reading "1. adding pagination for the results from the API".

Never mind the previous comments. Anyway, it has pagination feature and it can handle the maximum number of records, 500. It seems a good place to get nextPageUrl during parsing API results, and it locates in 
_fetch_resource method.

            if 200 <= response.status_code < 300:
                rslt_items = response.json()
                return_items.append(rslt_items.get("items"))
                next_url = rslt_items.get("pageDetails").get("nextPageUrl")

                if next_url:
                    retry_counter['count'] = 0
                    _fetch_resource(next_url, full_next_url=True,
                                    retry_counter=retry_counter,
                                    *args, **kwargs)